flow/src/main/java/com/sequenceiq/flow/service/FlowService.java

```diff
@@ -260,4 +271,66 @@ public Optional<OperationFlowsView> getLastFlowOperationByResourceCrn(String res
         List<FlowLog> flowLogs = flowLogDBService.getLatestFlowLogsByCrnInFlowChain(resourceCrn);
         return flowOperationStatisticsService.createOperationResponse(resourceCrn, flowLogs);
     }
+
+    public Page<FlowLogResponse> getFlowLogsByIds(List<String> flowIds, Pageable pageable) {
+        LOGGER.info("Getting flow logs by flow ids list {}", flowIds);
+        Page<FlowLog> flowLogResponses = flowLogDBService.getFlowLogsByFlowIdsCreatedDesc(new HashSet<>(flowIds), pageable);
+        return new PageImpl<>(flowLogResponses.stream().map(flowLog -> flowLogConverter.convert(flowLog)).collect(toList()),
+                flowLogResponses.getPageable(), flowLogResponses.getTotalElements());
+    }
+
+    public Page<FlowCheckResponse> getFlowChainsByChainIds(List<String> chainIds, Pageable pageable) {
+        LOGGER.info("Getting flow check responses by flow check ids list {}", chainIds);
+        Page<FlowChainLog> flowChainLogs = flowChainLogService
+                .findAllByFlowChainIdInOrderByCreatedDesc(new HashSet<>(chainIds), pageable);
+        Map<String, FlowChainLog> flowChainsMap = flowChainLogs
+                .stream().filter(flowChainLog -> flowChainLog.getParentFlowChainId() == null)
+                .collect(Collectors.toMap(FlowChainLog::getFlowChainId, Function.identity()));
+
+        List<FlowChainLog> allRelatedFlowChainsList = new ArrayList<>();
+        Map<String, String> parentChainIdByChainIdMap = new HashMap<>();
+        flowChainsMap.forEach((parentFlowChainId, parentFlowChainLog) -> {
+            List<FlowChainLog> relatedFlowChainLogs = flowChainLogService.getRelatedFlowChainLogs(List.of(parentFlowChainLog));
+            relatedFlowChainLogs.forEach(log -> parentChainIdByChainIdMap.put(log.getFlowChainId(), parentFlowChainId));
+            allRelatedFlowChainsList.addAll(relatedFlowChainLogs);
+            allRelatedFlowChainsList.add(parentFlowChainLog);
+        });
+
+        Set<String> flowChainIdsSet = allRelatedFlowChainsList.stream().map(FlowChainLog::getFlowChainId).collect(toSet());
+
+        Map<String, List<FlowLogWithoutPayload>> flowLogsByParentChainId = fetchFlowLogsByParentChainId(flowChainIdsSet,
+                parentChainIdByChainIdMap);
+
+        return new PageImpl<>(flowChainsMap.keySet().stream().map(parentFlowChainId ->
+                convertFlowChainToFlowCheckResponse(flowChainsMap.get(parentFlowChainId),
+                flowLogsByParentChainId.getOrDefault(parentFlowChainId, List.of()))).collect(toList()),
+                flowChainLogs.getPageable(), flowChainLogs.getTotalElements());
```

Comment:
this would be much more readable after splitting it like:
```
        List<FlowCheckResponse> result = flowChainsMap.keySet().stream()
                .map(parentFlowChainId -> convertFlowChainToFlowCheckResponse(flowChainsMap.get(parentFlowChainId),
                        flowLogsByParentChainId.getOrDefault(parentFlowChainId, List.of())))
                .collect(toList());
        return new PageImpl<>(result, flowChainLogs.getPageable(), flowChainLogs.getTotalElements());